src/core/net/netif.cpp

```diff
@@ -535,5 +544,97 @@ bool Netif::IsUnicastAddress(const Address &aAddress) const
     return rval;
 }
 
+otError Netif::AddInternalUnicastAddress(const NetifUnicastAddress &aAddress)
+{
+    otError              error = OT_ERROR_NONE;
+    NetifUnicastAddress *entry;
+    const size_t         num = OT_ARRAY_LENGTH(mUnicastAddressPool);
+
+    VerifyOrExit(!aAddress.GetAddress().IsLinkLocal(), error = OT_ERROR_INVALID_ARGS);
+
+    for (entry = mUnicastAddresses; entry; entry = entry->GetNext())
+    {
+        if (entry->GetAddress() == aAddress.GetAddress())
+        {
+            VerifyOrExit(entry >= &mUnicastAddressPool[0] && entry < &mUnicastAddressPool[mInternalUnicastsNumber],
+                         error = OT_ERROR_INVALID_ARGS);
+
+            entry->mPrefixLength = aAddress.mPrefixLength;
+            entry->mPreferred    = aAddress.mPreferred;
+            entry->mValid        = aAddress.mValid;
+            ExitNow();
+        }
+    }
+
+    VerifyOrExit(mInternalUnicastsNumber + mExternalUnicastsNumber < num, error = OT_ERROR_NO_BUFS);
+
+    entry = &mUnicastAddressPool[mInternalUnicastsNumber++];
+
+    // Copy the new address into the available entry and insert it in linked-list.
+    *entry            = aAddress;
+    entry->mNext      = mUnicastAddresses;
+    mUnicastAddresses = entry;
+
+    GetNotifier().Signal(OT_CHANGED_IP6_ADDRESS_ADDED);
+
+exit:
+    return error;
+}
+
+otError Netif::RemoveInternalUnicastAddress(const Address &aAddress)
+{
+    otError              error  = OT_ERROR_NONE;
+    NetifUnicastAddress *prev   = NULL;
+    NetifUnicastAddress *target = NULL;
+    NetifUnicastAddress *entry;
+
+    VerifyOrExit(mInternalUnicastsNumber > 0, error = OT_ERROR_NOT_FOUND);
+
+    // Find target to remove
+    for (entry = mUnicastAddresses; entry; entry = entry->GetNext())
+    {
+        if (entry->GetAddress() == aAddress)
+        {
+            VerifyOrExit(entry >= &mUnicastAddressPool[0] && entry < &mUnicastAddressPool[mInternalUnicastsNumber],
+                         error = OT_ERROR_INVALID_ARGS);
+
+            target = entry;
+            break;
+        }
+    }
+
+    VerifyOrExit(target != NULL, error = OT_ERROR_NOT_FOUND);
+
+    --mInternalUnicastsNumber;
+
+    // Find previous entry of the top entry of internal addresses stack
+    for (entry = mUnicastAddresses; entry != &mUnicastAddressPool[mInternalUnicastsNumber]; entry = entry->GetNext())
+    {
+        prev = entry;
+    }
+
+    // Remove the top entry from list
+    if (prev == NULL)
+    {
+        mUnicastAddresses = entry->GetNext();
+    }
+    else
+    {
+        prev->mNext = entry->GetNext();
+    }
+
+    // Replace value of target with value of top entry
+    if (target != entry)
+    {
+        entry->mNext = target->GetNext();
+        *target      = *entry;
+    }
```

Comment:
This can potentially change the order of addresses in the linked-list.
I think it's fine and order does not matter. @jwhui do we care about the order?